listenerutil/parse.go

```diff
@@ -387,3 +413,153 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
+	}
+
+	for _, crh := range customResponseHeader {
+		for statusCode, responseHeader := range crh {
+			headerValList, ok := responseHeader.([]map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
+			}
+
+			status, err := convertStatusCode(statusCode)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(headerValList) != 1 {
+				return nil, fmt.Errorf("invalid number of response headers exist")
+			}
+			headerValMap := headerValList[0]
+			headerVal, err := parseHeaders(headerValMap)
+			if err != nil {
+				return nil, err
+			}
+
+			h[status] = headerVal
+		}
+	}
+
+	// setting default headers
+	if h[0] == nil {
+		h[0] = make(map[string]string)
+	}
+	if _, ok := h[0]["Strict-Transport-Security"]; !ok {
+		h[0]["Strict-Transport-Security"] = strictTransportSecurity
+	}
+	if _, ok := h[0]["X-Content-Type-Options"]; !ok {
+		h[0]["X-Content-Type-Options"] = xContentTypeOptions
+	}
+	if _, ok := h[0]["Cache-Control"]; !ok {
+		h[0]["Cache-Control"] = cacheControl
+	}
+	if _, ok := h[0]["Content-Security-Policy"]; !ok {
+		if uiHeaders {
+			h[0]["Content-Security-Policy"] = uiContentSecurityPolicy
+		} else {
+			h[0]["Content-Security-Policy"] = apiContentSecurityPolicy
+		}
+	}
+
+	return h, nil
+}
+
+// isValidStatusCode checks for status codes outside the allowed range
+func convertStatusCode(sc string) (int, error) {
+	if sc == "default" {
+		return 0, nil
+	}
+	status, err := strconv.Atoi(sc)
+	if err != nil {
+		if _, err = fmt.Sscanf(sc, "%dxx", &status); err != nil {
+			return -1, fmt.Errorf("status does not match expected format. should be a valid status code or formatted \"%%dxx\". was: %s", sc)
+		}
+		if status > 5 || status < 1 {
+			return -1, fmt.Errorf("status is not within valid range, must be between 1xx and 5xx. was: %s", sc)
+		}
+		return status, nil
+	}
+	if status >= 600 || status < 100 {
+		return -1, fmt.Errorf("status is not within valid range, must be between 100 and 599. was: %s", sc)
+	}
+
+	return status, nil
+}
+
+func parseHeaders(in map[string]interface{}) (map[string]string, error) {
+	hvMap := make(map[string]string)
+	for k, v := range in {
+		// parsing header name
+		headerName := textproto.CanonicalMIMEHeaderKey(k)
+		// parsing header values
+		s, err := parseHeaderValues(v)
+		if err != nil {
+			return nil, err
+		}
+		hvMap[headerName] = s
+	}
+	return hvMap, nil
+}
+
+func parseHeaderValues(header interface{}) (string, error) {
+	var sl []string
+	headerValList, ok := header.([]interface{})
+	if !ok {
+		return "", fmt.Errorf("headers must be given in a list of strings")
+	}
+	for _, vh := range headerValList {
+		if _, ok := vh.(string); !ok {
+			return "", fmt.Errorf("found a non-string header value: %v", vh)
+		}
+		headerVal := strings.TrimSpace(vh.(string))
+		if headerVal == "" {
+			continue
+		}
+		sl = append(sl, headerVal)
+
+	}
+	s := strings.Join(sl, "; ")
```

Comment:
Instead of joining the strings like this, can we not keep the separate values as a slice and just add all of them with `Header().Add(v)`? This feels a bit ugly to be doing here.

Reply:
I don't think either of those are good enough reasons to keep this TBH. We should operate on these as well typed values until we know we want to merge them, and at that point we shouldn't be doing it manually with `;`, we should use the library tools (`w.Header().Add`), IMO.